lib/SIL/IR/SILFunctionType.cpp

```diff
@@ -1252,6 +1252,49 @@ enum class ConventionsKind : uint8_t {
   CXXMethod = 8,
 };
 
+template <typename T>
+std::optional<T>
+matchSwiftAttr(const clang::Decl *decl,
+               llvm::ArrayRef<std::pair<llvm::StringRef, T>> patterns) {
+  if (!decl || !decl->hasAttrs())
+    return std::nullopt;
+
+  for (const auto *attr : decl->getAttrs()) {
+    if (const auto *swiftAttr = llvm::dyn_cast<clang::SwiftAttrAttr>(attr)) {
+      for (const auto &p : patterns) {
+        if (swiftAttr->getAttribute() == p.first)
+          return p.second;
+      }
+    }
+  }
+  return std::nullopt;
+}
+
+template <typename T>
+std::optional<T> matchSwiftAttrConsideringInheritance(
+    const clang::Decl *decl,
+    llvm::ArrayRef<std::pair<llvm::StringRef, T>> patterns) {
+  if (!decl)
+    return std::nullopt;
+
+  if (auto match = matchSwiftAttr<T>(decl, patterns))
+    return match;
+
+  if (const auto *recordDecl = llvm::dyn_cast<clang::CXXRecordDecl>(decl)) {
+    for (const auto &baseSpecifier : recordDecl->bases()) {
+      const clang::CXXRecordDecl *baseDecl =
```

Comment:
nit: you should be able to do `const auto *baseDecl` here, right? It's already clear what the type of `baseDecl` is from the name of the method that returns it (i.e., `getAsCXXRecordDecl()`).